Python/optimizer_bytecodes.c

```diff
@@ -956,14 +956,48 @@ dummy_func(void) {
             // isinstance(inst, cls) where both inst and cls have
             // known types, meaning we can deduce either True or False
 
-            // The below check is equivalent to PyObject_TypeCheck(inst, cls)
             PyObject *out = Py_False;
             if (inst_type == cls_o || PyType_IsSubtype(inst_type, cls_o)) {
                 out = Py_True;
             }
             sym_set_const(res, out);
             REPLACE_OP(this_instr, _POP_CALL_TWO_LOAD_CONST_INLINE_BORROW, 0, (uintptr_t)out);
         }
+        else if (inst_type && sym_matches_type(cls, &PyTuple_Type)) {
+            // isinstance(inst, tup) where inst has a known type and tup is a tuple.
+            // We can deduce True if inst is an instance of at least one of
+            // the items in the tuple.
+            // We can deduce False if all items in the tuple have known types and
+            // inst is not an instance of any of them.
+
+            int length = sym_tuple_length(cls);
+            if (length != -1) {
+                // We cannot do anything about tuples with unknown length
+                PyObject *out = Py_False;
+                for (int i = 0; i < length; i++) {
+                    JitOptRef item = sym_tuple_getitem(ctx, cls, i);
+                    if (!sym_has_type(item)) {
+                        // There is an unknown item in the tuple.
+                        // It could potentially define its own __instancecheck__
+                        // method so we can only deduce bool.
+                        out = NULL;
+                        break;
+                    }
+                    PyTypeObject *cls_o = (PyTypeObject *)sym_get_const(ctx, item);
+                    if (cls_o &&
+                        sym_matches_type(item, &PyType_Type) &&
```

Comment:
Maybe add a comment explaining that this is to protect against metaclasses definine `__instancecheck__`.

Reply:
We're not only checking that the object is a subclass of `type`, we're also checking that it is an exact instance of `type` itself. We care about this second condition because it guarantees that `__instancecheck__` doesn't exist (otherwise we would need to look it up to check if it exists or not).

Reply:
Thanks for the clarification, I added a comment :)